### ISSUE TITLE
Double revision upon single file save (0.7)

### ISSUE THREAD

**Issue by yurivish** [timholy/Revise.jl#83]
I have a directory with the following two files:

```julia
# test.jl
function __main__()
    println("revise is...")
end
```

```julia
# watch.jl
ENV["JULIA_REVISE"] = "manual"
using Revise, FileWatching

file = joinpath(@__DIR__, "test.jl")

include(file)
Revise.track(file)

__main__()

while true
    event = watch_file(file)
    if event.changed
        revise()
        println("file changed; re-evaluating __main__.")
        @eval __main__()
    end
end
```

The idea is to support re-running a computation whenever a script or its dependencies have changed. 

I notice the following behavior when I run Julia from the terminal, wait for the script to start listening for changes, and then update `test.jl` to print `"revise is... magic"`, then save the file:

```
$ julia watch.jl
revise is...
file changed; re-evaluating __main__.
revise is...
file changed; re-evaluating __main__.
revise is... magic
```

For some reason, `revise is...` is printed twice: the second copy is printed immediately once I save the file, with the `magic` version following it about a second after.

Julia version info:

```
julia> versioninfo()
Julia Version 0.7.0-DEV.3476
Commit 625923f993 (2018-01-19 15:54 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.7.0)
  CPU: Intel(R) Core(TM) i7-3615QM CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
Environment:
  JULIA_EDITOR = subl
```

I'm going to pull the latest master now to see if the issue goes away. Update: Same issue on today's master:

```
Julia Version 0.7.0-DEV.3704
Commit f37d8d7a60 (2018-02-03 03:17 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.7.0)
  CPU: Intel(R) Core(TM) i7-3615QM CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
Environment:
  JULIA_EDITOR = subl
```

**Comment by timholy** [comment]
Your test case led me to discover a different bug---thanks!

I can only rarely reproduce this. I'm guessing it's some kind of near-race condition, and may depend on details like whether these files reside on a SSD or HDD. Moreover, I have an easier time reproducing it if I exclude Revise entirely:
```julia
using FileWatching

file = joinpath(@__DIR__, "test.jl")

include(file)

__main__()

counter = 0

while true
    event = watch_file(file)
    if event.changed
        counter += 1
        @show counter
        println("file changed; re-evaluating (but not revising) __main__.")
        @eval __main__()
    end
end
```
and fail to say `using FileWatching` (something about the deprecation machinery triggers it more reliably).

What happens if you add a short `sleep` inside the loop?

**Comment by yurivish** [comment]
If I add a short `sleep` at the top of the loop or in between the `revise()` and `@eval` it doesn't change anything — I always get output as if it runs the old `__main__`, then the current `__main__` around a second later.

Here's output with exact timings using `Dates.now()`:

```
2018-02-04T15:15:25.277 revise is...
file changed; re-evaluating __main__.
2018-02-04T15:15:30.505 revise is...
file changed; re-evaluating __main__.
2018-02-04T15:15:31.237 revise is... magic.
```
I thought for a while this might be a bug with Sublime Text, and so tried Vim, where upon I got an actual error:

```
2018-02-04T15:15:31.237 revise is... magic.
ERROR: LoadError: start_watching (File Monitor): no such file or directory (ENOENT)
Stacktrace:
 [1] uv_error at ./libuv.jl:68 [inlined]
 [2] start_watching(::FileMonitor) at /Users/yurivish/Documents/julia/usr/share/julia/site/v0.7/FileWatching/src/FileWatching.jl:355
 [3] wait(::FileMonitor) at /Users/yurivish/Documents/julia/usr/share/julia/site/v0.7/FileWatching/src/FileWatching.jl:433
 [4] watch_file(::String, ::Int64) at /Users/yurivish/Documents/julia/usr/share/julia/site/v0.7/FileWatching/src/FileWatching.jl:512
 [5] watch_file(::String) at /Users/yurivish/Documents/julia/usr/share/julia/site/v0.7/FileWatching/src/FileWatching.jl:492
 [6] top-level scope at /Users/yurivish/Dropbox/Projects/Julia Guide/guide/src/hs/watch.jl:12
 [7] include at ./boot.jl:296 [inlined]
 [8] include_relative(::Module, ::String) at ./loading.jl:1012
 [9] include(::Module, ::String) at ./sysimg.jl:26
 [10] exec_options(::Base.JLOptions) at ./client.jl:332
 [11] _start() at ./client.jl:447
in expression starting at /Users/yurivish/Dropbox/Projects/Julia Guide/guide/src/hs/watch.jl:12
```

Probably unrelated; maybe Vim temporarily removes the file while saving. 

Tried TextEdit... Same double-revise as Sublime. The problem persists even if I quit Dropbox (which I thought might be doing something with file metadata that would trigger the "changed" event).

Hrm.

**Comment by yurivish** [comment]
Ignoring the bug for a moment, is what I'm doing a generally reasonable approach to having a "re-run this computation with new definitions on every file save" workflow outside the REPL?

**Comment by timholy** [comment]
Sure, this is a reasonable thing to want to do.

What happens when you try my "Revise-free" demo above?

**Comment by yurivish** [comment]
If I run `julia watch2.jl` (containing the Revise-free demo), wait for it to start up, then modify and save the test script, I see the following output:

```
$ julia watch2.jl
revise is...
┌ Warning: Deprecated syntax `implicit assignment to global variable `counter``.
│ Use `global counter` instead.
└ @ nothing none:0
counter = 1
file changed; re-evaluating (but not revising) __main__.
revise is...
counter = 2
file changed; re-evaluating (but not revising) __main__.
revise is...
```
The `counter = 1` outputs print out immediately upon saving, with the `counter = 2` outputs following along around a second later.

**Comment by timholy** [comment]
Thanks. So the implication is that this isn't a Revise bug, it's a bug with that `watch_file` loop. Previous issues (#15, #22) have made me wonder if there's something slightly strange about OSX file-watching. I don't have a suitable machine to test on, unfortunately; perhaps worth reporting as a Julia bug? Or poke around a bit in `stdlib/FileWatching` and see if you can come up with some insight?

**Comment by yurivish** [comment]
And that happens even for the following script:

```julia
using FileWatching

file = joinpath(@__DIR__, "test2.jl")

counter = 0

while true
    event = watch_file(file)
    if event.changed
        counter += 1
        @show counter
        println("file changed; re-evaluating (but not revising) __main__.")
    end
end
```

**Comment by yurivish** [comment]
Yeah, I'll poke around, thanks.

There might still be a Revise issue, since the initial "changed" event doesn't seem to properly update the code run by the next `@eval __main__()`, which seems to execute with the old code. I used `@eval` to circumvent any potential world-age issues, but there's still some reason why it isn't picking up the changes even after sleeping for a short while. Does that sound like a Revise bug to you?

**Comment by yurivish** [comment]
Observation: The double-change only occurs when the file has content — if it's empty, then I just get the initial change event, and nothing further. Might imply something about the way either Sublime Text or the OS writes/flushes files... I'll open an issue about it on the main Julia repo. 😄

**Comment by yurivish** [comment]
Another data point: Run the following script with a file called `file.txt` in the same directory. The contents of every file change are reflected immediately in the printed messages, rather than only in the second message (as with Revise).

```
using FileWatching

file = joinpath(@__DIR__, "file.txt")
counter = 0

while true
    event = watch_file(file)
    if event.changed
        global counter += 1
        @show counter
        println("file changed: $(read(file, String)).")
    end
end
```

**Comment by timholy** [comment]
> There might still be a Revise issue, since the initial "changed" event doesn't seem to properly update the code run by the next @eval __main__(), which seems to execute with the old code.

That may be a timing/scheduler issue: if the file hasn't entered the [revision queue](https://github.com/timholy/Revise.jl/blob/4942692a7cc88b2a5bcc5716d3c1db6186674f01/src/Revise.jl#L35), the call to `revise()` won't do anything. The queue gets populated by [watching directories and then scanning their `mtime`s](https://github.com/timholy/Revise.jl/blob/4942692a7cc88b2a5bcc5716d3c1db6186674f01/src/pkgs.jl#L86-L98). If the event in your loop gets processed before the event in Revise, then it will run the old code.

You should be able to get it to guarantee new code by sleeping (say, `sleep(0.1)`) before your call to `revise()`. 

> and so tried Vim, where upon I got an actual error

Your error you triggered with Vim (which you'll note was in your loop, not in Revise) is also familiar; Revise works around this, if the file/directory is unexpectedly missing [these lines](https://github.com/timholy/Revise.jl/blob/4942692a7cc88b2a5bcc5716d3c1db6186674f01/src/Revise.jl#L186-L191) give it a second chance. I'd triggered it most recently by stashing/unstashing in git, but I remember someone reporting the same problem with Vim.

Overall, it's a little tricky writing asynchronous code that interacts with the file system!

**Comment by yurivish** [comment]
> You should be able to get it to guarantee new code by sleeping (say, sleep(0.1)) before your call to revise().

Ooh, that works. Awesome! Thank you! Interestingly trying the same thing with `yield()` in place of sleep or using `@schedule` did not suffice. I tried those thinking that maybe the other tasks were already queued but my file-watching got unblocked first. 

Yes, tricky business. I wonder if there's a more principled way to address these sorts of issues other than taking strategic naps...

**Comment by timholy** [comment]
I've been wondering the same thing. If you're comfortable monitoring an internal structure, perhaps you could `yield` until `Revise.revision_queue` stops growing. But it seems a little risky, because any other tasks might make it look like Revise is done growing the queue.

**Comment by timholy** [comment]
Other than for the Julia bug, OK to close this?

**Comment by yurivish** [comment]
Yes, I think so. Thanks again for your help — I now have a wonderfully interactive workflow set up for the project I'm working on.

**Comment by timholy** [comment]
Great! I'd be happy to accept a README PR if you want to advertise this solution.
